sw/device/tests/usbdev_logging_test.c

```diff
@@ -0,0 +1,98 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// USB logging test
+
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/lib/testing/usb_logging.h"
+
+// Presently we must reinstate the UART output manually.
+#include "sw/device/lib/dif/dif_uart.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+/**
+ * Number of concurrent logging streams to test.
+ */
+static const unsigned kNumStreams = 4u;
+/**
+ * Number of iterations of logging message.
+ */
+static const unsigned kNumIters = 1000u;
+
+OTTF_DEFINE_TEST_CONFIG();
+
+bool test_main(void) {
+  LOG_INFO("Running USBDEV_LOGGING test");
+
+  // ----------------- Test the more generic, direct interface ----------------
+
+  // Enable USB logging
+  CHECK_STATUS_OK(
+      usb_logging_init(NULL, 0u, kNumStreams, (1u << kNumStreams) - 1u));
+
+  for (uint8_t s = 0u; s < kNumStreams; s++) {
+    usb_logging_text(s, "USB - Logging via direct interface\n");
+  }
+
+  for (unsigned iter = 0u; iter < kNumIters; iter++) {
+    for (uint8_t s = 0u; s < kNumStreams; s++) {
+      char buf[44];
+      size_t len = base_snprintf(buf, sizeof(buf),
+                                 "S%u: Now logging %u over USB\n", s, iter);
+      CHECK_STATUS_OK(usb_logging_data(s, (uint8_t *)buf, len));
+    }
+  }
+
+  // Finalize the logging code.
+  CHECK_STATUS_OK(usb_logging_fin(true));
+
+  // ------------------- Test redirection of stdout via OTTF ------------------
+
+  // Enable logging of stdout traffic over USB
+  CHECK_STATUS_OK(usb_logging_enable());
+
+  for (unsigned iter = 0u; iter < kNumIters; iter++) {
+    LOG_INFO("Now logging %u over USB", iter);
+  }
+
+  // TODO: just a quantity of data as ASCII hex...
+  if (true) {
+    base_hexdump((char *)0x20000000, 128 * 1024);
+  }
+
+  CHECK_STATUS_OK(usb_logging_fin(true));
+
+  // TODO: this is the only way that presently we can reinstate the UART
+  // and thus report test completion...only - though I do not know why - this
+  // actually DOES permit logging output again, but often does not result in
+  // test completion.
+  //
+  // It seems to be very sensitive to the logging output that was sent before
+  // the following code; it is NOT even necessary to perform USB logging or
+  // to switch the stdout other than the call below to our DIF instance.
+
+  dif_uart_t uart0;
+  CHECK_DIF_OK(dif_uart_init(
```

Comment:
@timothytrippel would you know why the UART needs to be initialized again?

Reply:
I've resolved this; reinitializing the UART via another DIF instance sort of worked, but exhibited malfunction that was very sensitive to the test behavior; it's not an appropriate solution was only an interim. I've discovered that I needed to modify the OTTF config to set 'test_may_clobber' in the console settings. This properly reinitialized the console-over-UART output, emits the test status and opentitantool responds appropriately.